### ISSUE TITLE
8$: Warn on for(;;){}while(cond);

### ISSUE THREAD

**Issue by strager** [quick-lint/quick-lint-js#954]
In the following examples, the programmer intended to write a do-while loop:
```javascript
for(;;) {
} while (cond);

while (cond) {
} while (cond);
```

If we see `while` after a `for` or `while` loop's body, without any newline after the `}`, report a diagnostic.

**Comment by footgod368** [comment]
I claim this for-hire task. I expect payment after I complete this task. I will email the quick-lint-js team if I am assigned this task.

In addition, I'm an junior student in University of Michigan, and one project of my Software Engineering course requires me to contribute to an open-source codebase.

**Comment by cphamWSU** [comment]
Hello, I'd like to take on this task as part of classwork to contribute to an open source project.

This is my first time doing so, and I'd love to reach out for some guidance. What would be the best medium to ask for help?

**Comment by strager** [comment]
> Hello, I'd like to take on this task as part of classwork to contribute to an open source project.

Awesome! I'm happy to help you with this.

> What would be the best medium to ask for help?

Here are [some options](https://quick-lint-js.com/contact/):

* This GitHub issue
* My Twitch stream: https://twitch.tv/strager
* The quick-lint-js IRC channel: [#quick-lint-js on LiberaChat](https://kiwiirc.com/nextclient/irc.libera.chat/#quick-lint-js)
* My email: strager.nds@gmail.com

**Comment by cphamWSU** [comment]
Thanks for the quick response. I've been familiarizing myself with the relevant code and implementing the initial steps for a diagnostic. Now that I'm onto writing the production code, I wanted to check if I was on the right track.

Using `parse_and_visit_do_while` from `parse-statement.cpp` as a reference for how to evaluate source code tokens, my assumption is that the solution for this issue would be to add some logic that checks for tokens following the right curly brace inside of `parse_and_visit_for` and `parse_and_visit_while` in the same file. I would imagine that `has_leading_newline` will also end up playing a role.

Does any of that sound right?

**Comment by strager** [comment]
@cphamWSU Adding checks at the end of `parse_and_visit_for` sounds about right. Checking `has_leading_newline` for the `while` token sounds like a good idea too.
